presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java

```diff
@@ -0,0 +1,252 @@
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import static io.prestosql.plugin.phoenix.PhoenixErrorCode.PHOENIX_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class PhoenixSplitManager
+        implements ConnectorSplitManager
+{
+    private static final Logger log = Logger.get(PhoenixSplitManager.class);
+
+    private final PhoenixClient phoenixClient;
+    private PhoenixMetadata phoenixMetadata;
+    private final Map<String, HostAddress> hostCache = new HashMap<>();
+
+    @Inject
+    public PhoenixSplitManager(PhoenixClient phoenixClient, PhoenixMetadataFactory phoenixMetadataFactory)
+    {
+        this.phoenixClient = requireNonNull(phoenixClient, "client is null");
+        this.phoenixMetadata = phoenixMetadataFactory.create();
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorTableLayoutHandle layout, SplitSchedulingStrategy splitSchedulingStrategy)
+    {
+        PhoenixTableLayoutHandle layoutHandle = (PhoenixTableLayoutHandle) layout;
+        PhoenixTableHandle handle = layoutHandle.getTable();
+        try (PhoenixConnection connection = phoenixClient.getConnection()) {
+            String inputQuery = phoenixClient.buildSql(
+                    connection,
+                    handle.getCatalogName(),
+                    handle.getSchemaName(),
+                    handle.getTableName(),
+                    layoutHandle.getDesiredColumns(),
+                    layoutHandle.getTupleDomain(),
+                    phoenixMetadata.getColumns(handle, false));
+
+            return new FixedSplitSource(getHadoopInputSplits(inputQuery).stream().map(split -> (PhoenixInputSplit) split).map(split -> {
+                List<HostAddress> addresses = getSplitAddresses(split);
+
+                return new PhoenixSplit(
+                        phoenixClient.getConnectorId(),
+                        handle.getCatalogName(),
+                        handle.getSchemaName(),
+                        handle.getTableName(),
+                        layoutHandle.getTupleDomain(),
+                        addresses,
+                        new WrappedPhoenixInputSplit(split));
+            }).collect(Collectors.toList()));
+        }
+        catch (IOException | InterruptedException | SQLException e) {
+            throw new PrestoException(PHOENIX_ERROR, e);
+        }
+    }
+
+    private List<HostAddress> getSplitAddresses(PhoenixInputSplit split)
+    {
+        List<HostAddress> addresses;
+        try {
+            String hostName = split.getLocations()[0];
+            HostAddress address = hostCache.get(hostName);
```

Comment:
This can be
```java
HostAddress address = hostCache.computeIfAbsent(hostName, HostAddress::fromString);

Reply:
I'm not convinced we need a cache at all here. `HostAddress.fromString()` is doing basic parsing to split a string into a host and port. There is no address resolution or anything else expensive.